// @microsoft/fast-element/dist/esm/index.js
export { FAST, emptyArray } from "./platform.js";
// DOM
export { DOMAspect, DOM } from "./dom.js";
// Observation
export { ExecutionContext, Observable, observable, SourceLifetime, volatile, } from "./observation/observable.js";
export { PropertyChangeNotifier, SubscriberSet, } from "./observation/notifier.js";
export { ArrayObserver, lengthOf, Splice, SpliceStrategy, SpliceStrategySupport, } from "./observation/arrays.js";
export { Updates } from "./observation/update-queue.js";
// Binding
export { Binding } from "./binding/binding.js";
export { listener, oneWay } from "./binding/one-way.js";
export { oneTime } from "./binding/one-time.js";
export { normalizeBinding } from "./binding/normalize.js";
// Styles
export { ElementStyles, } from "./styles/element-styles.js";
export { css } from "./styles/css.js";
export { cssDirective, CSSDirective, } from "./styles/css-directive.js";
export {} from "./styles/host.js";
export {} from "./styles/style-strategy.js";
export { CSSBindingDirective } from "./styles/css-binding-directive.js";
// Templating
export { html, InlineTemplateDirective, ViewTemplate, } from "./templating/template.js";
export { Compiler } from "./templating/compiler.js";
export { Markup, Parser } from "./templating/markup.js";
export { HTMLBindingDirective, } from "./templating/html-binding-directive.js";
export { htmlDirective, HTMLDirective, StatelessAttachedAttributeDirective, } from "./templating/html-directive.js";
export { ref, RefDirective } from "./templating/ref.js";
export { when } from "./templating/when.js";
export { repeat, RepeatBehavior, RepeatDirective, } from "./templating/repeat.js";
export { slotted, SlottedDirective, } from "./templating/slotted.js";
export { children, ChildrenDirective, } from "./templating/children.js";
export { HTMLView } from "./templating/view.js";
export { elements, NodeObservationDirective, } from "./templating/node-observation.js";
// Components
export { customElement, FASTElement } from "./components/fast-element.js";
export { FASTElementDefinition, } from "./components/fast-definitions.js";
export { attr, AttributeConfiguration, AttributeDefinition, booleanConverter, nullableBooleanConverter, nullableNumberConverter, } from "./components/attributes.js";
export { ElementController, } from "./components/element-controller.js";
